### PR TITLE
[FLINK-10566] Fix exponential planning time of large programs

### PR DESCRIPTION
**Pull request by mxm**
## What is the purpose of the change

The traversal of the DAG is not efficient enough at some places which can lead
to very long plan creation times.

This introduces caching for the traversal to avoid traversing nodes multiple
times. 

## Brief change log

Caching is performed at two places:

- when registering Kryo types
- when determining the maximum parallelism

## Verifying this change

This change added tests and can be verified as follows:

Run `LargePlanTest` test to verify.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: yes
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no

### CODE REVIEW

File: flink-tests/src/test/java/org/apache/flink/test/planning/LargePlanTest.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+				public String map(String value) throws Exception {
+					return value;
+				}
+				}).withBroadcastSet(stats.map(s -> "(" + s + ").map"), "stats");
+			}
+
+			TupleTypeInfo<Tuple2<Integer, String>> typeInfo =
+				new TupleTypeInfo<>(TypeInformation.of(Integer.class), TypeInformation.of(String.class));
+			DataSet<String> branch = input
+				.map(s -> new Tuple2<>(0, s + ii)).returns(typeInfo)
```
Comment (fhueske) [inline flink-tests/src/test/java/org/apache/flink/test/planning/LargePlanTest.java]: can be simplified to `.returns(Types.TUPLE(Types.STRING, Types.INT))`

File: flink-tests/src/test/java/org/apache/flink/test/planning/LargePlanTest.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+	public void testPlanningOfLargePlan() throws Exception {
+		runProgram(new PreviewPlanEnvironment(), 10, 50);
+	}
+
+	private static void runProgram(ExecutionEnvironment env, int depth, int width) throws Exception {
+		DataSet<String> input = env.fromElements("a", "b", "c");
+		DataSet<String> stats = null;
+
+		for (int i = 0; i < depth; i++) {
+		stats = analyze(input, stats, width / (i + 1) + 1);
```
Comment (fhueske) [inline flink-tests/src/test/java/org/apache/flink/test/planning/LargePlanTest.java]: indent

File: flink-tests/src/test/java/org/apache/flink/test/planning/LargePlanTest.java [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+				}).withBroadcastSet(stats.map(s -> "(" + s + ").map"), "stats");
+			}
+
+			TupleTypeInfo<Tuple2<Integer, String>> typeInfo =
+				new TupleTypeInfo<>(TypeInformation.of(Integer.class), TypeInformation.of(String.class));
+			DataSet<String> branch = input
+				.map(s -> new Tuple2<>(0, s + ii)).returns(typeInfo)
+				.groupBy(0)
+				.minBy(1)
+				.map(kv -> kv.f1).returns(TypeInformation.of(String.class));
```
Comment (fhueske) [inline flink-tests/src/test/java/org/apache/flink/test/planning/LargePlanTest.java]: `returns(Types.STRING)`

**Comment by mxm** [comment]
Thanks for the review @fhueske. I've pushed the style changes. I'll rerun the tests before merging.

File: flink-tests/src/test/java/org/apache/flink/test/planning/LargePlanTest.java [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+import org.junit.Test;
+
+import java.util.Collection;
+
+/**
+ * Tests that large programs can be compiled to a Plan in reasonable amount of time.
+ */
+public class LargePlanTest {
+
+	@Test(expected = OptimizerPlanEnvironment.ProgramAbortException.class, timeout = 15_000)
```
Comment (zentol) [inline flink-tests/src/test/java/org/apache/flink/test/planning/LargePlanTest.java]: What is the expected runtime of this test on travis & locally?

File: flink-tests/src/test/java/org/apache/flink/test/planning/LargePlanTest.java [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+	private static DataSet<String> analyze(DataSet<String> input, DataSet<String> stats, int branches) {
+		for (int i = 0; i < branches; i++) {
+			final int ii = i;
+
+			if (stats != null) {
+				input = input.map(new RichMapFunction<String, String>() {
+
+				@Override
+				public void open(Configuration parameters) throws Exception {
+					Collection<String> broadcastSet = getRuntimeContext().getBroadcastVariable("stats");
```
Comment (zentol) [inline flink-tests/src/test/java/org/apache/flink/test/planning/LargePlanTest.java]: is this actually relevant to the test?

Comment (mxm) [reply flink-tests/src/test/java/org/apache/flink/test/planning/LargePlanTest.java]: No, it's not relevant for the test. Will remove.

File: flink-core/src/main/java/org/apache/flink/api/common/Plan.java [hunk 6]
```diff
@@ -362,9 +363,13 @@ public int getMaximumParallelism() {
 	private static final class MaxDopVisitor implements Visitor<Operator<?>> {
 
 		private int maxDop = -1;
+		private Set<Operator> visitedOperators = new HashSet<>();
```
[the remark this answers is not in the document]
Comment (mxm) [reply flink-core/src/main/java/org/apache/flink/api/common/Plan.java]: Yes, will make it final.

File: flink-tests/src/test/java/org/apache/flink/test/planning/LargePlanTest.java (hunk 4 above)
Comment (mxm) [reply flink-tests/src/test/java/org/apache/flink/test/planning/LargePlanTest.java]: On my machine this runs in 500ms. Without the fix this would runs for hours or never return.

**Comment by mxm** [comment]
Btw, I also want to backport this to 1.5 and 1.6. Do you know if there is another 1.5 release planned? Or is 1.5 already end of life with 1.7 being out?

**Comment by fhueske** [comment]
There was a discussion about a 1.5.6 on the dev ML (started by Till) a few days ago and it was concluded that there will be another release.

**Comment by mxm** [comment]
That's great. I'm running into timeouts in the connectors build: https://travis-ci.org/apache/flink/builds/467099321

Failures look unrelated and the tests ran fine on my Travis account. Rerunning just to make sure.

**Comment by zentol** [comment]
FYI: If you restart a Travis build the previous build (i.e., the one with the timeout) is no longer accessible to anyone.

**Comment by mxm** [comment]
That's odd. Good to know. There were some fixes pushed to the connectors yesterday. Let's see what the rerun shows.

**Comment by mxm** [comment]
All green this time. Merging.
